File: 0001-0500/0011. Container With Most Water.java
// Medium
// Two Pointers
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/container-with-most-water/

class Solution {
    public int maxArea(int[] height) {
        int left = 0, right = height.length - 1, maxArea = 0;
        while (left < right) {
            int minHeight = Math.min(height[left], height[right]);
            int curArea = minHeight * (right - left);
            maxArea = Math.max(maxArea, curArea);

            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }
        return maxArea;
    }
}

File: 0001-0500/0011. Container With Most Water.py
# Time: O(n), Space:O(1)

from typing import *

class Solution:
    def maxArea(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        max_water = 0

        while left < right:
            w = right - left
            h = min(height[left], height[right])
            max_water = max(max_water, w * h)

            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
        
        return max_water
    